### ISSUE TITLE
Document better keybindings in the README

### ISSUE THREAD

**Issue by minad** [oantolin/embark#237]
Just some idea, but I kind of love this now. I have bound `embark-default-action` to `M-.` and `embark-act` to `C-.`. This makes a lot of sense since `xref-find-definitions` is bound to `M-.` and Embark generalizes this. It may also make sense to rename `embark-default-action` to `embark-act-default` for the symmetry.

**Comment by hmelman** [comment]
Very interesting.  I think the command `embark-default-action` and how the default action is defined (is it by binding it to RET?) should be mentioned in the README 😄

**Comment by oantolin** [comment]
@minad  These do seem like sensible key bindings. Well, for graphical Emacs: I don't think `C-.` works on many terminals (`M-.` should be fine in terminals). I guess I already have an example of some keybinding, I might as well have it be a nicer one. (Maybe what I'd really like is to not have any key binding, but people seem to like the ability to copy and paste a configuration.) 

On a personal note, I'm wondering if I should bind `embark-default-action` now, I do seem to use the default action a lot more these days...

@hmelman Good point! I will do that. I tend to regard the docstrings as the primary documentation and forget the manual a bit. 🙄

**Comment by oantolin** [comment]
Oh, and to answer your question @hmelman, as the `embark-default-action` docstring says, the default action is determined as follows:

First the `embark-default-action-overrides` alist associating target types to default actions is consulted. Failing that, if you are in the minibuffer, the default action is the command you ran that opened the minibuffer; if you aren't in the minibuffer, the RET binding in the relevant action keymap is used (I call the RET binding the "default default action" in my head).

The idea of the overrides alist is that the current system says that, for example, if you run `M-x delete-file` the default action is `delete-file`, which makes sense to me, but someone might prefer that `find-file` always be the default action for files, no matter what command is prompting for a file.

The default configuration specifies no overrides and embark-consult uses one for an interesting reason: `consult-grep` candidates have the funny property that they are not valid input for `consult-grep` usually, and if they are they might not match themselves! I guess you could say `consult-grep` is really presenting a menu of options, not _completing_ candidates. So embark-consult sets an overriding default action for the `consult-grep` type that goes to the indicated location. Notice that this simulates what the minibuffer rule usually accomplishes: that using the default action is like picking that completion candidate and pressing RET.

**Comment by minad** [comment]
You are right about the terminal. But you could add a comment that on terminal one might want to use sth different.

In my case the keybinding was more accidental. I used C-. all along and then it occured to me that I should bind default-action too (which should be called embark-act-default for symmetry) and M-. just fits perfectly then.

I didn't consider alternatives, but I really do like that it shadows the existing binding - one key less to learn. To me Embark feels like a natural extension here.

**Comment by oantolin** [comment]
This is off-topic, but `C-.` reminded me of what I use it for: `dot-mode-execute`, and I want to do a little propaganda for the great dot-mode package. It's an even better version of vi/vim's `.` command, which repeats the last editing command (i.e., it doesn't repeat motion commands or buffer switching commands, etc., just editing commands). What dot-mode does is similar but even better: instead of repeating the last editing command, it repeats the last consecutive stretch of editing commands. In Vim I often stopped to think, "Is there a way to do this edit with a single command so I can repeat it or do I need to record a macro?" Using dot-mode is like always having the foresight of recording a keyboard macro before you do some edits. (And if you do need one or two non-editing commands, say motion commands, to be included in the commands repeated you just need to prefix them with `dot-mode-override`.)

**Comment by oantolin** [comment]
Oh, `C-S-a` which is currently mentioned in the README doesn't work in my terminal either! I think I'll suggest `C-.`, `C-;` and `M-E` for terminal users.

**Comment by minad** [comment]
You could also document multiple keybindings, this would underline that users should make their own choices. Your choice is certainly fine for the use-package example. My suggestion is a bit more intrusive, since it uses the `xref-find-defintions` -> `embark-default-action` generalization at point. Works well so far :)

What do you think about renaming `embark-default-action` to `embark-act-default`? I think this makes sense, since the default action is also supposed to be used as a toplevel command and the symmetry in the name helps the understanding for new users. It is a bit like "google search" and "I'm feeling lucky".

**Comment by oantolin** [comment]
I am proposing to document multiple keybindings!

There's two very minor things I don't like about the name `embark-act-default`: (1) it's not the current name, so I'd have to deprecate and in the future remove the old one 😛, (2) it's not grammatical (I know this probably shouldn't bother me...). I don't think the name `embark-default-action` is confusing in any way.

**Comment by hmelman** [comment]
I do think of it as `embark-I-feel-lucky` 😆 While (1) is always an issue I think that (2) also applies to `embark-default-action` since that sounds like a noun and not a verb particularly when compared with `embark-act` and `embark-become`.  I do like a name that relates it with `embark-act`.  Maybe something like `embark-act-do-default` or  `embark-act-with-default`?

**Comment by minad** [comment]
> I don't think the name embark-default-action is confusing in any way.

It is not confusing, but it is weird that `embark-act` and `embark-default-action` are commands on an equal footing. I think the idea of having only a single main entry point to Embark should be given up :-P

It is like left and right click.

**Comment by oantolin** [comment]
How about `embark-dwim` for the default action? (I'm warming up to `embark-act-default`, but still want to consider some alternatives.) Or `embark-act-now!`?

**Comment by hmelman** [comment]
If those are the choices I prefer `embark-dwim` over `embark-act-now!` and definitely wouldn't like an `!` in the name :)

**Comment by oantolin** [comment]
The `!` was tongue in cheek. What do you think about `embark-act-now`?

**Comment by hmelman** [comment]
As I said, I prefer -dwim over it and I think all the other options so far too.  -dwim is now a common convention for emacs commands so that people seeing that name will understand it does something immediately given the situation.  That's very fitting here and given embark's general features I suspect most people will intuit that it just does the default action.  It's easy to see a good docstring would making it complete and quite discoverable.

embark-act-now has the advantage of obviously relating to embark-act.  I'd find the -now a little vaguer as to intent.  I mean all commands act when you invoke them so it's not different from embark-act in that respect ;)  It's usually delays that are mentioned in command names like `run-at-time`.

I think embark-dwim is understandable (at least guessable) given just it's name.  I think embark-act-now requires the context of embark-act to be understood.  IMHO.

**Comment by minad** [comment]
I prefer embark-dwim over embark-act-now. I am not a big fan of dwim in command names, but in this case it feels okay.

**Comment by oantolin** [comment]
OK, `embark-act-now` is out of the running. Between `embark-dwim` and `embark-act-default` which do you prefer?

**Comment by oantolin** [comment]
And what about between `embark-dwim` and `embark-act-default`?

**Comment by hmelman** [comment]
I prefer `embark-dwim` over `embark-act-default`

**Comment by minad** [comment]
`embark-dwim` is more to the point.

**Comment by oantolin** [comment]
So `embark-dwim` it is! @minad kindly made a pull request that implements these changes, but I am having second doubts about the key bindings and wanted to summarize some possibilities here:

I do want to suggest some better key bindings in the documentation (or rather, I don't want to suggest any keybindings since I think that's a very personal matter, but I also recognize the convenience of having a copy-pasteable configuration even though I never copy and paste configurations myself), but I am unsure which key bindings to suggest. 

Ideally the key bindings suggested shouldn't shadow any built-in key bindings of either Emacs or any of the completion UIs Embark works with, and ideally they should also work in most terminal emulators.

For embark-act we have:

- `C-S-a`, the current suggestion. I don't think anyone likes this (I certainly don't).
- `C-.`, @minad's suggestion. I like it a lot (though wouldn't use it myself, because I prefer to use it for dot-mode), but icomplete does use it to cycle among completions, which is pretty important, I guess.
- `C-;`, what I use. As far as I can tell, this doesn't clash with anything.
- `C-,`, what @protesilaos uses. It clashes with `org-cycle-agenda-files`, which I expect is a fairly popular command. That command is also bound to `C-'`, so maybe this is OK.

I think none of this work in the terminal. :(

For embark-dwim we have:
- none! `<embark-act> RET` is not so terrible.
- `M-.`, @minad's proposal. Again, I like this a lot, but it does clash with `xref-find-definitions` and while it does do the same thing on a symbol, as @minad points out, users might still want to run `xref-find-definitions` directly some times (I often use it with a prefix argument, for example).
- `C-,`, what I am using currently. Again org binds this to `org-cycle-agenda-files`
- `M-RET`, this is what Hyperbole uses for the action key. This clashes with Hyperbole, of course, but also with Org and with completion UIs such as Vertico and Selectrum which use it to submit the minibuffer contents as they are (except they wimp out if `require-match` is `t` :P). That's probably too many clashes, right?

I think `M-.` and `M-RET` work in the terminal, as does "none", if you choose a binidng for `embark-act` which works in the terminal.

**Comment by oantolin** [comment]
Also, probably the binding of `embark-act` in Embark collect buffers (which is currently `a`) should change to match whatever the readme recommends. If we go with `C-.`, say, we could change it to `.`. But maybe this isn't important, since you can definitely use your global bindings in the collect buffers too.

**Comment by protesilaos** [comment]
On 2021-06-13, 10:56 -0700, Omar Antolín Camarena ***@***.***> wrote:

> So `embark-dwim` it is! @minad kindly made a pull request that
> implements these changes, but I am having second doubts about the key
> bindings and wanted to summarize some possibilities here:

Haven't followed this thread.  Will check later.

> - `C-,`, what @protesilaos uses. It clashes with
> `org-cycle-agenda-files`, which I expect is a fairly popular
> command. That command is also bound to `C-'`, so maybe this is OK.

Actually the C-, is the legacy of the old exit/no-exit scheme of
embark-act.  I had the C-, bound to the no-exit variant and the C-. to
the exiting one, kind of like how the comma and the dot work for
sentences.  Eventually got used to C-, so I stayed with that.

> I think none of this work in the terminal. :(
>
> For embark-dwim we have:
> - none! `<embark-act> RET` is not so terrible.
> - `M-.`, @minad's proposal. Again, I like this a lot, but it does
> clash with `xref-find-definitions` and while it does do the same thing
> on a symbol, as @minad points out, users might still want to run
> `xref-find-definitions` directly some times (I often use it with a
> prefix argument, for example).
> - `C-,`, what I am using currently. Again org binds this to
> `org-cycle-agenda-files`
> - `M-RET`, this is what Hyperbole uses for the action key. This
> clashes with Hyperbole, of course, but also with Org and with
> completion UIs such as Vertico and Selectrum which use it to submit
> the minibuffer contents as they are (except they wimp out if
> `require-match` is `t` :P). That's probably too many clashes, right?
>
> I think `M-.` and `M-RET` work in the terminal, as does "none", if you
> choose a binidng for `embark-act` which works in the terminal.

How about the keys reserved for users, like F5..F9?  They should be
safe, given this is a recommendation, and they work in the terminal.

-- 
Protesilaos Stavrou
https://protesilaos.com

**Comment by minad** [comment]
This question of the recommended binding is not important, since Embark does not grab a binding by default. I mostly care about having a better name `embark-dwim` and having `embark-dwim` documented in the README!

I think it is okay to recommend opinionated bindings which even shadow some reserved Emacs bindings (my C-. and M-. suggestion). Generally the keybinding allocation is not well thought out upstream. I am not happy that almost every C/M keybinding is preallocated and only bindings, which are not well accessible are made available for custom bindings, at least as long as one does not use additional modifiers like super.

**Comment by hmelman** [comment]
FWIW I'm on a mac and have the apple command key configured as super key in emacs (I only use GUI), I'm not sure how convenient it is to make a super key on other systems but having one is a big win for emacs bindings.  I'm doing:

``` emacs-lisp
(global-set-key (kbd "s-;") 'embark-act)
(global-set-key (kbd "M-s-;") 'embark-dwim)
```

If you make your example config the above, you have the advantage of not conflicting with builtin or other package bindings :) Though in some seriousness it lets someone copy a config to their `init.el` and find out quickly that there are two bindings for them to pick and that if they can configure a super key, their emacs might get easier to use.

I like using semi-colon for `embark-act` because the command feels to me like taking the previous thing (clause) and moving on to another related thing (clause).  Or combining two commands in a shell command line. Adding another modifier to that for `embark-dwim` fits my mental model well.

Of your above suggestions I like C-; and M-.  I think org is popular enough you don't want to conflict with it and I don't think Hyperbole meets that threshold.

**Comment by oantolin** [comment]
I merged @minad's PR #241 which renames `embark-default-action` to `embark-dwim` and suggests `C-.` for `embark-act` and `C-;` for `embark-dwim`. I'll further update the docs to mention the other possibilities and to explain what the default action is.
